cloud_governance/cloud_resource_orchestration/aws/ec2/monitor_tickets.py

```diff
@@ -0,0 +1,221 @@
+            es_id_data = self.__es_operations.get_es_data_by_id(id=ticket_id, index=self.es_cro_index).get('_source')
+            if es_id_data:
+                ticket_opened_date = description.get('TicketOpenedDate').date()
+                self.__region_name = description.get('Region')
+                ticket_monitoring_days = (current_date - ticket_opened_date).days
+                duration = int(es_id_data.get('duration'))
+                remaining_duration = duration - ticket_monitoring_days
+                es_data_change = self.verify_es_instances_state(es_data=es_id_data)
+                if es_data_change:
+                    self.__es_operations.update_elasticsearch_index(index=self.es_cro_index, id=ticket_id, metadata=es_id_data)
+                self.__alert_in_progress_ticket_users(ticket_id=ticket_id, es_id_data=es_id_data, remaining_duration=remaining_duration)
+
+    @typeguard.typechecked
+    @logger_time_stamp
+    def __tag_extend_instances(self, sub_tasks: list, ticket_id: str, duration: int, sub_task_count: int, estimated_cost: float):
+        """
+        This method extend the duration if the user opened expand ticket
+        :param sub_tasks:
+        :param ticket_id:
+        :param duration:
+        :param sub_task_count:
+        :param estimated_cost:
+        :return:
+        """
+        ticket_id = ticket_id.split('-')[-1]
+        local_ec2_operations = EC2Operations(region=self.__region_name)
+        local_ec2_client = boto3.client('ec2', region_name=self.__region_name)
+        filters = {'Filters': [{'Name': f'tag:{self.__cro_resource_tag_name}', 'Values': [ticket_id]}]}
+        extend_duration = 0
+        extended_cost_estimation = 0
+        for task_id in sub_tasks:
+            description = self.__jira_operations.get_issue_description(ticket_id=task_id, sub_task=True)
+            extend_duration += int(description.get('Days'))
+            extended_cost_estimation += float(description.get('CostEstimation'))
+        instance_ids = local_ec2_operations.get_ec2_instance_ids(**filters)
+        if duration > 0:
+            duration += extend_duration
+            estimated_cost += extended_cost_estimation
+            tags = [{'Key': 'Duration', 'Value': str(duration)}, {'Key': 'EstimatedCost', 'Value': str(round(estimated_cost, self.DEFAULT_ROUND_DIGITS))}]
+            local_ec2_operations.tag_ec2_resources(client_method=local_ec2_client.create_tags, resource_ids=instance_ids, tags=tags)
+            data = {'duration': duration, 'timestamp': datetime.utcnow(), 'sub_tasks': len(sub_tasks) + sub_task_count, 'estimated_cost': round(estimated_cost, self.DEFAULT_ROUND_DIGITS)}
+            if self.__es_operations:
+                self.__es_operations.update_elasticsearch_index(metadata=data, id=ticket_id, index=self.es_cro_index)
+            for task_id in sub_tasks:
+                self.__jira_operations.move_issue_state(ticket_id=task_id, state='closed')
```

Comment:
Closed ticket when duration>0, is that OK ?

Reply:
Subtasks are closed after the updation of cost and days.